src/Elliptic/Actions/InitializeAnalyticSolution.hpp

```diff
@@ -32,47 +32,53 @@ namespace Actions {
 /*!
  * \brief Place the analytic solution of the system fields in the DataBox.
  *
+ * If the `BackgroundTag` holds an `AnalyticSolutionType`, evaluate it for all
+ * `AnalyticSolutionFields` and store the fields in the DataBox. Otherwise,
+ * signal no analytic solution is available by storing `std::nullopt` in the
+ * DataBox.
+ *
  * Uses:
  * - DataBox:
- *   - `AnalyticSolutionTag`
+ *   - `BackgroundTag`
  *   - `Tags::Coordinates<Dim, Frame::Inertial>`
  *
  * DataBox:
  * - Adds:
- *   - `db::wrap_tags_in<::Tags::Analytic, AnalyticSolutionFields>`
+ *   - `::Tags::AnalyticSolutions<AnalyticSolutionFields>`
  *
  * \note This action relies on the `SetupDataBox` aggregated initialization
  * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
  * phase action list prior to this action.
  */
-template <typename AnalyticSolutionTag, typename AnalyticSolutionFields>
+template <typename BackgroundTag, typename AnalyticSolutionFields,
+          typename AnalyticSolutionType = typename BackgroundTag::type>
```

Comment:
In that case the code won't actually compile if the type isn't an analytic solution, because the compiler tries to instantiate its `variables` function and fails if it doesn't provide the required tags. And if the class provides the tags then it might as well be treated as an analytic solution :)

Reply:
Okay, but I thought that kind of 'loose' interface specification (where it doesn't matter the inheritance path, as long as it defines the right functions it 'works') was precisely what you were attempting to avoid by specifying base classes / protocols?

Reply:
Yea in fact the type passed as background here might not actually be an analytic solution, even if it provides the required tags - it could just be an initial guess with data for all fields, but that doesn't necessarily solve the equations. I'll have to remove the default parameter in a separate commit though or squash the two commits together with the one upgrading the executables, because else the commit won't compile anymore.